### DISCUSSION TITLE
Change upgrade behaviour

### DISCUSSION THREAD

**Discussion by davidmashworth** [Romanitho/Winget-AutoUpdate#561] [General]
Hi all,

Apologies if this has already been discussed. I have just deployed this via Intune and testing with a small group of users. What we have noticed is that during updates, old versions of apps are left behind. I understand in some cases this may be required for keeping certain versions etc... but in our case, we would like to just update to the current version and remove all older versions.

From some reading, I have seen you can potentially use "upgradeBehavior: uninstallPrevious" - is that right? If so how would I go about implementing this?

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/77258962/adadee41-90d3-4a82-83b9-110c67167363)

Interestingly some apps like Notepad++ did the upgrade and no old versions remained

**Comment by AndrewDemski-ad-gmail-com** [comment]
Hi @davidmashworth,
You may be observing something what was discussed in #539

WAU is a decision-making layer on top of a winget tool which relies on data provided via repository.
That repository is feed by software manufacturers (quality varies).
If manifests are inconsistent, then winget makes mistakes.

### I wet through these commands to reproduce your problem.

winget install Oracle.JavaRuntimeEnvironment --version 8.0.3210.7
>Found Java 8 [Oracle.JavaRuntimeEnvironment] Version 8.0.3210.7

winget list Oracle.JavaRuntimeEnvironment

| Name | Id | Version | Available | Source|
|-----|-----|-----|-----|-----|
|Java 8 Update 321 (64-bit) |Oracle.JavaRuntimeEnvironment| 8.0.3210.7| 8.0.4010.10 |winget|

winget install Oracle.JavaRuntimeEnvironment
>Found an existing package already installed. Trying to upgrade the installed package...
>Found Java 8 [Oracle.JavaRuntimeEnvironment] Version 8.0.4010.10

I ended up with one java, the latest one.

We operate based on several preliminary assumptions:
- all software is distributed at the system level
- manufacturers provide manifests that allow WINGET to remove all versions
- installation packages from INTUNE will not force the installation of an older version due to the old baseline defined in intune

Could any of the above have influenced the presented effect?

**Reply by davidmashworth** [reply]
Hi @AndrewDemski-ad-gmail-com ,

Thanks for your reply. I think it may be a combination of issues. Sorry in advance for my questions below

If we just take Java as an example, here are the logs:

`10:53:43 - Updating Java 8 Update 321 (64-bit) from 8.0.3210.7 to 8.0.4010.10...
10:53:47 - ##########   WINGET UPGRADE PROCESS STARTS FOR APPLICATION ID 'Oracle.JavaRuntimeEnvironment'   ##########
10:53:47 - -> Running: Winget upgrade --id Oracle.JavaRuntimeEnvironment --accept-package-agreements --accept-source-agreements -h
 
   - 

 
  ███▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  1024 KB / 9.20 MB
  ██████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  2.00 MB / 9.20 MB
  █████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  3.00 MB / 9.20 MB
  █████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  4.00 MB / 9.20 MB
  ████████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒  5.00 MB / 9.20 MB
  ███████████████████▒▒▒▒▒▒▒▒▒▒▒  6.00 MB / 9.20 MB
  ██████████████████████▒▒▒▒▒▒▒▒  7.00 MB / 9.20 MB
  ██████████████████████████▒▒▒▒  8.00 MB / 9.20 MB
  █████████████████████████████▒  9.00 MB / 9.20 MB
  ██████████████████████████████  9.20 MB / 9.20 MB
No installed package found matching input criteria.
10:53:50 - ##########   WINGET UPGRADE PROCESS FINISHED FOR APPLICATION ID 'Oracle.JavaRuntimeEnvironment'   ##########
10:53:50 - Java 8 Update 321 (64-bit) updated to 8.0.4010.10 !
10:53:54 - Updating Java 8 Update 321 from 8.0.3210.7 to 8.0.4010.10...
10:53:58 - ##########   WINGET UPGRADE PROCESS STARTS FOR APPLICATION ID 'Oracle.JavaRuntimeEnvironment'   ##########
10:53:58 - -> Running: Winget upgrade --id Oracle.JavaRuntimeEnvironment --accept-package-agreements --accept-source-agreements -h
 
   - 

 
  ███▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  1024 KB / 9.20 MB
  ██████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  2.00 MB / 9.20 MB
  █████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  3.00 MB / 9.20 MB
  █████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  4.00 MB / 9.20 MB
  ████████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒  5.00 MB / 9.20 MB
  ███████████████████▒▒▒▒▒▒▒▒▒▒▒  6.00 MB / 9.20 MB
  ██████████████████████▒▒▒▒▒▒▒▒  7.00 MB / 9.20 MB
  ██████████████████████████▒▒▒▒  8.00 MB / 9.20 MB
  █████████████████████████████▒  9.00 MB / 9.20 MB
  ██████████████████████████████  9.20 MB / 9.20 MB
No installed package found matching input criteria.
10:54:00 - ##########   WINGET UPGRADE PROCESS FINISHED FOR APPLICATION ID 'Oracle.JavaRuntimeEnvironment'   ##########
10:54:00 - Java 8 Update 321 updated to 8.0.4010.10 !`

In terms of how Intune handles the install

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/77258962/248d28e3-35e5-4ce1-8989-0ab9417e511f)

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/77258962/647486b7-df0b-4682-ad2b-246373d7ca39)

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/77258962/70a0f8d1-de19-4be5-803d-be7c35cb9905)

From the images above, is this what you were thinking re your last point on how Intune forces the installation of an older version? How would we change this behaviour?

Thanks,
David

**Reply by AndrewDemski-ad-gmail-com** [reply]
Devils in the details, what is set in the detection rules/methods?
If Intune doesn't repeatedly install the current JRE from the current baseline (independently/outside of the update cycle initiated by WAU), then the detection rule must be valid and returns a condition representing status=present for this product (nothing to do).
Sometimes this detection is related to the presence of a specific folder, file or files with a specific file version.
If the upgrade replaces the entire product and detection is based on these file-based conditions, Intune will assume that the product is missing and must be reinstalled. then you will get two javas. 
1. Intune will install current JRE,
2. WAU will trigger winget upgrade (which removes JRE from pt.1) 
3. Package detection rule for that Intune package indicates state=missing and Intune will reinstall older JRE (next to newest).

**Reply by davidmashworth** [reply]
Detection method uses the MSI product code

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/77258962/dfaf1648-596e-4cf8-b885-583eac97864c)

**Reply by AndrewDemski-ad-gmail-com** [reply]
OK, 
and what is going to happen to state of that ProductCode when winget performs upgrade and removes older version?
:)

**Reply by Romanitho** [reply]
I always use a file detection method. Check for the Java EXE app in the program files and check if the version is greater or equal to the version you deploy. This way, you're always compliant is WAU is executed ;)

**Reply by Romanitho** [reply]
And for your 7zip, I bet for a mix of EXE or MSI installers :)
